apt: enable uca repositories

gboutry opened the pull request:
- [X] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?
  * `lint-pyright` fails, no logs about why

-----
Example of rockcraft.yaml:
```yaml
name: keystone
summary: Openstack Keystone
license: Apache-2.0
description: |
  Ubuntu distribution of OpenStack Keystone
version: antelope

base: ubuntu:22.04
platforms:
  amd64:

package-repositories:
  - type: apt
    cloud: antelope
    pocket: updates
    priority: always

parts:
  keystone:
    plugin: nil
    stage-packages:
      - apache2
      - keystone
```

javacruft commented on craft_archives/repo/package_repository.py:
```diff
@@ -27,6 +27,26 @@
+    "xena": "focal",
+    "yoga": "focal",
+    "zed": "jammy",
+    "antelope": "jammy",
```
we can forward plan this a bit and add "bobcat": "jammy" as well

gboutry replied:
Removed the `RELEASE_MAP` dict

gboutry commented on craft_archives/repo/apt_sources_manager.py:
```diff
@@ -98,6 +98,7 @@ def _install_sources(
         suites: List[str],
         url: str,
         keyring_path: pathlib.Path,
+        ignore_keyring_path: bool = False,
```
It does, but we need to install the ubuntu-cloud-keyring.

Installation of the keyring id defined here:
https://github.com/gboutry/craft-archives/blob/295dc072d3bce60654667a640b64015b13c9fa7b/craft_archives/repo/apt_uca.py#L27

And used in conjunction with ignore_keyring: 
https://github.com/gboutry/craft-archives/blob/295dc072d3bce60654667a640b64015b13c9fa7b/craft_archives/repo/apt_sources_manager.py#L248

But this is mostly due to errors in testing, I will try to find a better way (this was already bothering me and is one of the main reasons it's a draft for the moment)

tigarmo replied:
(fixed by reworking the logic to use the keyserver)

tigarmo commented:
Thanks for the work! Before I get on the nitty-gritty of the code review I think we should discuss the `ubuntu-cloud-keyring` package thing. There are two issues with that:
1) This extra dependency on an archive package means calling `apt install` "out of band" from the rest of the code; Specifically, as far as I can tell currently only `craft-parts` manipulates the repository so we would have to check the consequences of this change;
2) More importantly (and I think this is what @sergiusens was referring to), this `ubuntu-cloud-keyring` installs the keyring in [/etc/apt/trusted.gpg.d/](https://packages.ubuntu.com/jammy/all/ubuntu-cloud-keyring/filelist), which is the behavior we moved away from; keyrings are installed in `/etc/apt/keyrings` and then referenced specifically in the repo's source definition.

I would very much like to have the cloud-archive repo be consistent with the other repo types (key in /etc/apt/keyrings, referenced by source.list file in /etc/apt/sources.list.d/). I think we could support that if the keyring is available in keyserver.ubuntu.com, @gboutry do you know if the key is there?

gboutry commented:
Yes, it looks like the key is present at http://keyserver.ubuntu.com/pks/lookup?search=391A9AA2147192839E9DB0315EDB1B62EC4926EA&fingerprint=on&op=index

gboutry commented:
In last fp:
- UCA keyring no longer installed, using the KEY_ID to fetch the key from the key server

tigarmo reviewed (commented):
This is very nice! I think the only thing left is, for sanity's sake, to test this in rockcraft itself (via a spread test)
I can do it for you but I'll need a valid `cloud` repo and the name of a package it contains; do you have these handy?
Thanks again!

gboutry commented:
> This is very nice! I think the only thing left is, for sanity's sake, to test this in rockcraft itself (via a spread test) I can do it for you but I'll need a valid `cloud` repo and the name of a package it contains; do you have these handy? Thanks again!

Thanks @tigarmo, of course.

In `antelope`, you will be able to find:
- `keystone` at `23.0.0-0ubuntu1~cloud0`
- `nova-compute` at `27.0.0-0ubuntu1~cloud0`

In `zed`, you will be able to find:
- `keystone` at `22.0.0-0ubuntu1~cloud0`
- `nova-compute` at `26.1.0-0ubuntu1~cloud0`

Note that `antelope` and `zed` are the only cloud archive compatible with `jammy` at the moment. If you want to add tests for `focal` for example, you will need to test with either `yoga`, `xena`, `wallaby` or `victoria`

tigarmo commented:
Thanks! Here's a test case that failed in rockcraft but is missing in craft-archives (missing from `test_projects.py`):

```python
def test_validate_repository_empty_dict():
    with pytest.raises(pydantic.ValidationError):
        validate_repository({})
```

gboutry commented:
> Thanks! Here's a test case that failed in rockcraft but is missing in craft-archives (missing from `test_projects.py`):
> 
> ```python
> def test_validate_repository_empty_dict():
>     with pytest.raises(pydantic.ValidationError):
>         validate_repository({})
> ```

Indeed, there was an issue in the control flow. I fixed it and added this test case!

tigarmo reviewed (approved):
The builds for rockcraft and snapcraft passed; I think the remaining missing coverage is fine because we should toss all of that duplicated "manual validation" soon anyway (there's an issue for it).
So LGTM! Thanks again!
